jni/src/test/java/com/horizen/sc2scnative/Sc2ScTest.java

```diff
@@ -0,0 +1,290 @@
+package com.horizen.sc2scnative;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.util.Optional;
+import java.util.Random;
+
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import com.horizen.TestUtils;
+import com.horizen.certnative.WithdrawalCertificate;
+import com.horizen.commitmenttreenative.CommitmentTree;
+import com.horizen.commitmenttreenative.CustomBitvectorElementsConfig;
+import com.horizen.commitmenttreenative.CustomFieldElementsConfig;
+import com.horizen.commitmenttreenative.ScCommitmentCertPath;
+import com.horizen.librustsidechains.Constants;
+import com.horizen.librustsidechains.FieldElement;
+import com.horizen.merkletreenative.InMemoryAppendOnlyMerkleTree;
+import com.horizen.merkletreenative.MerklePath;
+import com.horizen.provingsystemnative.ProvingSystem;
+import com.horizen.provingsystemnative.ProvingSystemType;
+
+public class Sc2ScTest {
+    private static int MAX_QUALITY_CERT_HASH_CUSTOM_FIELDS_POS = Constants.MAX_QUALITY_CERT_HASH_CUSTOM_FIELDS_POS();
+    private static int MSG_ROOT_HASH_CUSTOM_FIELDS_POS = Constants.MSG_ROOT_HASH_CUSTOM_FIELDS_POS();
+    private static int MIN_CUSTOM_FIELDS = Constants.MIN_CUSTOM_FIELDS();
+    private static int MSG_MT_HEIGHT = Constants.MSG_MT_HEIGHT();
+
+    static boolean zk = false;
+
+    static ProvingSystemType psType = ProvingSystemType.COBOUNDARY_MARLIN;
+    static int customFieldsNum = 3;
+    static String pkPath = "./test_sc2sc_pk";
+    static String vkPath = "./test_sc2sc_vk";
+
+    static int maxProofPlusVkSize = 9 * 1024;
+
+    private InMemoryAppendOnlyMerkleTree msgTree;
+    private CommitmentTree currentCt = CommitmentTree.init();
+    private CommitmentTree nextCt = CommitmentTree.init();
+    private WithdrawalCertificate currWithdrawalCertificate;
+    private WithdrawalCertificate nextWithdrawalCertificate;
+
+    @BeforeClass
+    public static void initKeys() throws Exception {
+        ProvingSystem.generateDLogKeys(psType, TestUtils.DLOG_KEYS_SIZE);
+        assertTrue(Sc2Sc.setup(psType, customFieldsNum, Optional.of(TestUtils.CERT_SEGMENT_SIZE),
+                pkPath, vkPath, zk, maxProofPlusVkSize));
+    }
+
+    @AfterClass
+    public static void deleteKeys() {
+        // Delete proving keys and verification keys
+        new File(pkPath).delete();
+        new File(vkPath).delete();
+    }
+
+    @Before
+    public void setUp() {
+        msgTree = InMemoryAppendOnlyMerkleTree.init(MSG_MT_HEIGHT, 1 << MSG_MT_HEIGHT);
```

Comment:
I think is safest to expose  a class to handle the message tree, otherwise we have to trust that the proof creator will construct the tree with this  correct parameters.
we could have a Sc2scMessageTree JNI class with methods for:
- add new message hash
- get merkle path from message hash 
- get root

Reply:
I've exposed that `MSG_MT_HEIGHT` constant in `Constants` class. That is the only business logic that we need.... I guess that the SDK can handle the message tree and the relative paths to fill correctly the custom fields and the requested path in the circuit creation.